contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,192 @@
+        require(amount != 0, "Transfer value of 0");
+
+        uint256 adjustedAmount = amount;
+        if (tokenReserve > 0) {
+            if (amount > tokenReserve) {
+                emit CollateralAdded({amount: tokenReserve, remaining: 0});
+                adjustedAmount = amount - tokenReserve;
+                tokenReserve = 0;
+            } else {
+                tokenReserve -= amount;
+                emit CollateralAdded({amount: amount, remaining: tokenReserve});
+                return;
+            }
+        }
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, adjustedAmount);
+        emit MintNativeTokens(recipient, adjustedAmount);
+    }
+
+    /**
+     * @dev See {INativeTokenMinter-bridgeTokens}.
+     */
+    function transferToSource(
+        address recipient,
+        address feeContractAddress,
+        uint256 feeAmount,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        require(
+            tokenReserve == 0,
+            "Cannot release tokens until source contract is collateralized"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedAmount = 0;
+        if (feeAmount > 0) {
+            adjustedAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeContractAddress),
+                feeAmount
+            );
+            IERC20(feeContractAddress).safeIncreaseAllowance(
+                address(teleporterMessenger),
+                adjustedAmount
+            );
+        }
+
+        // Burn native token by sending to BLACKHOLE_ADDRESS
+        payable(BLACKHOLE_ADDRESS).transfer(msg.value);
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationChainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: TeleporterFeeInfo({
+                    contractAddress: feeContractAddress,
+                    amount: adjustedAmount
+                }),
+                requiredGasLimit: TRANSFER_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, msg.value)
+            })
+        );
+
+        emit TransferToSource({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            feeContractAddress: feeContractAddress,
+            feeAmount: adjustedAmount,
+            teleporterMessageID: messageID
+        });
+    }
+}
```

Comment:
We discussed offline, but wanted to note here that we recommend adding an `isCollateralized` view method, and also a `totalSupply` view method that returns the total amount moved into the subnet minus the balance of the black hole fee address.